storage/metrics.go

```diff
@@ -0,0 +1,187 @@
+}
+
+func (s *UserMetrics) runMessageAggregation(query io.Reader) (*queryResult, error) {
+	res, err := s.es.Search(
+		s.es.Search.WithContext(context.Background()),
+		s.es.Search.WithIndex(s.indexName),
+		s.es.Search.WithBody(query),
+		s.es.Search.WithTrackTotalHits(true),
+		s.es.Search.WithPretty(),
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+	queryResult, err := unmarshalMessageAggregationQueryResult(res)
+	if err != nil {
+		return nil, err
+	}
+	return queryResult, nil
+}
+
+// This report returns # of messages by user in channel.
+// Additional grouping by day/hour/minute is possible.
+// If groupBy option is provided, the amount of returned data is limited according to constants in this module.
+func (s *UserMetrics) GetUserMessagesCounts(opts *api.UserActivityReportOptions) (*api.UserActivityReport, error) {
+	query := buildUserActivityQuery(opts)
+	query_result, err := s.runMessageAggregation(query)
+	if err != nil {
+		return nil, api.ErrInternal
+	}
+
+	results := make(api.UserActivityReport)
+	for _, hit := range query_result.Aggregations.Channels.Buckets {
+		activityItem := &api.ChannelActivity{TotalMessages: hit.DocCount}
+		if hit.BySpecifiedRange != nil {
+			activityItem.MessageCountDistribution = []api.DistributionEntry{}
+			for _, dateBucket := range hit.BySpecifiedRange.Buckets {
+				activityItem.MessageCountDistribution = append(activityItem.MessageCountDistribution, api.DistributionEntry{
+					PeriodStart:      dateBucket.KeyString,
+					MessagesInPeriod: dateBucket.DocCount,
+				})
+			}
+		}
+		results[hit.Key] = activityItem
+	}
+	return &results, nil
+}
+
+func buildUserActivityQuery(opts *api.UserActivityReportOptions) io.Reader {
+	var tf termFilter
+	tf.Term.Author = &termFilterValue{Value: opts.Author}
+
+	var aggs UserActivityAggs
+	aggs.Channels.Terms.Field = "channel.keyword"
+	aggs.Channels.Terms.Size = MaxChannelsInReport
+	aggs.Channels.Aggs = buildDateHistogramAggs(api.GroupBy(opts.GroupBy))
+
+	var q searchQuery
+	q.Query.Bool.Filter = []termFilter{tf}
+	q.Aggs = aggs
+
+	var b bytes.Buffer
+	json.NewEncoder(&b).Encode(q)
+	return &b
+}
+
+func buildDateHistogramAggs(groupBy api.GroupBy) *DateHistogramAggs {
+	if groupBy == "" {
+		return nil
+	}
+	var date_subaggs DateHistogramAggs
+	date_subaggs.BySpecifiedRange.DateHistogram.Field = "timestamp.as_date"
+	date_subaggs.BySpecifiedRange.DateHistogram.CalendarInterval = string(groupBy)
+	date_subaggs.BySpecifiedRange.DateHistogram.MinDocCount = 0
+	if groupBy == "day" {
+		date_subaggs.BySpecifiedRange.DateHistogram.Format = FormatDay
+	} else {
+		date_subaggs.BySpecifiedRange.DateHistogram.Format = FormatMinute
+	}
+
+	min_timestamp := getMinTimestamp(groupBy)
+	date_subaggs.BySpecifiedRange.DateHistogram.ExtendedBounds.Min = min_timestamp
+	date_subaggs.BySpecifiedRange.DateHistogram.ExtendedBounds.Max = time.Now().UnixMilli()
+	date_subaggs.BySpecifiedRange.DateHistogram.HardBounds.Min = min_timestamp
```

Comment:
Alternatively you could add `from` and `to` options in `UserActivityReportOptions` to define desired time range of the report

Reply:
done in last two commits